echopype/preprocess/api.py

```diff
@@ -329,3 +284,427 @@ def remove_noise(ds_Sv, ping_num, range_sample_num, noise_max=None, SNR_threshol
+    unique_er_bin_ind, unique_inverse = np.unique(digitized_echo_range, axis=0, return_inverse=True)
+
+    # create groups of row indices using the unique inverse
+    grps_same_ind = [
+        np.argwhere(unique_inverse == grp).flatten() for grp in np.unique(unique_inverse)
+    ]
+
+    # for each group bin and mean arr along echo_range
+    # note: the values appended may not be in the correct final order
+    binned_er = []
+    for count, grp in enumerate(grps_same_ind):
+        binned_er.append(
+            bin_and_mean_echo_range(arr[grp, :], unique_er_bin_ind[count, :], n_bin_er)
+        )
+
+    # construct er_means and put the columns in the correct order
+    binned_er_array = np.hstack(binned_er)
+    correct_column_ind = np.argsort(np.concatenate(grps_same_ind))
+    er_means = binned_er_array[:, correct_column_ind]
+
+    return er_means
+
+
+def bin_and_mean_2d(
+    arr: Union[dask.array.Array, np.ndarray],
+    bins_time: np.ndarray,
+    bins_er: np.ndarray,
+    times: np.ndarray,
+    echo_range: np.ndarray,
+    comp_er_check: bool = True,
+) -> np.ndarray:
+    """
+    Bins and means ``arr`` based on ``times`` and ``echo_range``,
+    and their corresponding bins. If ``arr`` is ``Sv`` then this
+    will compute the MVBS.
+
+    Parameters
+    ----------
+    arr: dask.array.Array or np.ndarray
+        The 2D array whose values should be binned
+    bins_time: np.ndarray
+        1D array (used by np.digitize) representing the binning required for ``times``
+    bins_er: np.ndarray
+        1D array (used by np.digitize) representing the binning required for ``echo_range``
+    times: np.ndarray
+        1D array corresponding to the time values that should be binned
+    echo_range: np.ndarray
+        2D array of echo range values
+    comp_er_check: bool
+        If True, a more comprehensive check will be completed to determine if ``echo_range``
+        grouping along ``ping_time`` is needed, otherwise a less comprehensive check will be done
+
+    Returns
+    -------
+    final_reduced: np.ndarray
+        The final binned and mean ``arr``, if ``arr`` is ``Sv`` then this is the MVBS
+
+    Notes
+    -----
+    This function assumes that ``arr`` has rows corresponding to
+    ``ping_time`` and columns corresponding to ``echo_range``.
+
+    This function should not be run if the number of ``echo_range`` values
+    vary amongst ``ping_times``.
+    """
+
+    # get the number of echo range and time bins
+    n_bin_er = len(bins_er)
+    n_bin_time = len(bins_time)
+
+    # obtain the bin indices for echo_range and times
+    digitized_echo_range, bin_time_ind = get_bin_indices(echo_range, bins_er, times, bins_time)
+
+    # determine if grouping of echo_range values with the same step size is necessary
+    if comp_er_check:
+        grouping_needed = is_grouping_needed_comprehensive(echo_range)
+    else:
+        grouping_needed = is_grouping_needed_less_comprehensive(echo_range)
```

Comment:
I have the following suggestions on refactoring/renaming the variables and functions in this block:
- rename `grouping_needed` to `er_grouping_need` just to make it more specific
- rename `comp_er_check` to `comprehensive_er_check` just to make it more specific
- rename the current 2 functions:
   - `is_grouping_needed_comprehensive` --> `if_all_er_steps_identical`
   - `is_grouping_needed_less_comprehensive` --> `if_last_er_steps_identical`
- combine the above 2 checking functions into one `is_er_grouping_needed` with `comp_er_check` as the input argument
   ```python
   def is_er_grouping_needed(echo_range, comprehensive_er_check):
      if comp_er_check:
         return if_all_er_steps_identical(echo_range)
      else:
         return if_last_er_steps_identical(echo_range)
   ```
My thought is that this way when reading this function, one can abstract away whatever the operations under the 2 checking functions are and just know that one is more comprehensive than the other, unless they are interested in knowing the details.

Reply:
All of these changes sound good to me! I will go ahead and implement them.